### PR TITLE
Default Network Policy for CIS

### PR DESCRIPTION
**Pull request by briandowns**
Adds a default network policy to the kube-system, default, and kube-public namespaces to adhere to CIS 1.5 compliance requirements. Addresses #208 

The default policy is set to allow only inter-namespace, pod to pod traffic.

Verification:

Check that the network policy has been created:

```sh
kubectl get networkpolicy
```

Check that annottions have been applied 

```sh
for i in "kubec-system kube-public default"; do 
    kubectl describe namespace ${i} | grep 'np.rke2.io: resolved'
done
```

Make sure that a deployed service isn't accessible outside of the namespace it's deployed to.

### CODE REVIEW

File: pkg/rke2/np.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	if err != nil {
+		return fmt.Errorf("networkPolicy: get %s - %w", namespace, err)
+	}
+	if ns.Annotations == nil {
+		ns.Annotations = make(map[string]string)
+	}
+	if _, ok := ns.Annotations[namespaceAnnotationNetworkPolicy]; !ok {
+		_, err := cs.NetworkingV1().NetworkPolicies(namespace).Get(ctx, defaultNetworkPolicyName, metav1.GetOptions{})
+		if err != nil {
+			if apierrors.IsAlreadyExists(err) {
```
Comment (dweomer) [inline pkg/rke2/np.go:58]: Not sure this branch will ever be followed for a GET.

File: pkg/rke2/np.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+				if err := cs.NetworkingV1().NetworkPolicies(namespace).Delete(ctx, defaultNetworkPolicyName, metav1.DeleteOptions{}); err != nil {
+					return err
+				}
+			} else if !apierrors.IsNotFound(err) {
+				return err
+			}
+		}
+		if _, err := cs.NetworkingV1().NetworkPolicies(namespace).Create(ctx, &networkPolicy, metav1.CreateOptions{}); err != nil {
```
Comment (dweomer) [inline pkg/rke2/np.go:66]: So the semantics here are intentional replacement instead of "patching" via UPDATE?

Comment (dweomer) [reply pkg/rke2/np.go:66]: To be clear, this is fine if that is what is required. A comment to that effect would be :+1:

File: pkg/rke2/psp.go [hunk 3]
```diff
@@ -324,16 +325,46 @@ func setPSPs(clx *cli.Context) func(context.Context, daemonsConfig.Control) erro
...
+					if apierrors.IsConflict(err) {
+						if err := updateNamespace(ctx, cs, ns); err != nil {
+							return err
+						}
+					}
+					return err
+				}
+				return nil
+			}); err != nil {
+				logrus.Fatalf("psp: update namespace: %s - %s", ns.Name, err.Error())
```
Comment (dweomer) [inline pkg/rke2/psp.go:339]: psp -> nsp ?

File: pkg/rke2/psp.go [hunk 4]
```diff
@@ -324,16 +325,46 @@ func setPSPs(clx *cli.Context) func(context.Context, daemonsConfig.Control) erro
...
 			logrus.Info("Applying PSP's complete")
 		}()
 		return nil
 	}
 }
 
+// updateNamespace receives a value of type v1.Namespace pointer
+// and updates that value to point to a newly retrieve value in
+// the event a conflict error is returned.
+func updateNamespace(ctx context.Context, cs *kubernetes.Clientset, ns *v1.Namespace) error {
```
Comment (dweomer) [inline pkg/rke2/psp.go:350]: with this func name i was expecting to see an `cs.CoreV1().Namespaces().Update()` invocation. can we rename to `getNamespace` or `cacheNamespace`, etc?

File: pkg/rke2/np.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+const (
+	namespaceAnnotationNetworkPolicy = "np.rke2.io"
+
+	defaultNetworkPolicyName = "default-network-policy"
+)
+
+// networkPolicy specifies a base level network policy applied
+// to the 3 primary namespace: kube-system, default, and kube-public.
+// This policy only allows for all inter-namespace traffic.
```
Comment (cjellick) [inline pkg/rke2/np.go:25]: You mean it only allows traffic within the namespace, right? If so, I think you want "intra-namespace", not "inter-namespace".

File: pkg/rke2/np.go (hunk 1 above)
Comment (cjellick) [reply pkg/rke2/np.go:58]: yeah, I think the delete needs to be outside of this if `err != nil` block all together and this `err != nil` should just return the error if it is any error other than `IsNotFound`.

File: pkg/rke2/np.go (hunk 2 above)
Comment (cjellick) [reply pkg/rke2/np.go:66]: agree on the comment. answering intention here: yes, intention is that if the annotation is not present, this is either a fresh cluster or the user is explicitly asking for a "reset" of net policy. So, just nuke and pave.

File: pkg/rke2/np.go (hunk 1 above)
Comment (briandowns) [reply pkg/rke2/np.go:58]: That's a lot cleaner.

File: pkg/rke2/psp.go [hunk 6]
```diff
@@ -324,16 +325,46 @@ func setPSPs(clx *cli.Context) func(context.Context, daemonsConfig.Control) erro
...
 		return nil
 	}
 }
 
+// updateNamespaceRef receives a value of type v1.Namespace pointer
+// and updates that value to point to a newly retrieve value in
+// the event a conflict error is returned.
+func updateNamespaceRef(ctx context.Context, cs *kubernetes.Clientset, ns *v1.Namespace) error {
+	logrus.Info("updating namespace: " + ns.Name)
+	newNS, err := cs.CoreV1().Namespaces().Get(ctx, metav1.NamespaceSystem, metav1.GetOptions{})
```
Comment (brandond) [inline pkg/rke2/psp.go:352]: Unless the intent is to update kube-system regardless of what namespace is passed in...
```suggestion
	newNS, err := cs.CoreV1().Namespaces().Get(ctx, ns.Name, metav1.GetOptions{})
```

File: pkg/rke2/np.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+// policy exists. If it does, we delete it, and create the new default policy.
+func setNetworkPolicy(ctx context.Context, namespace string, cs *kubernetes.Clientset) error {
+	ns, err := cs.CoreV1().Namespaces().Get(ctx, namespace, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("networkPolicy: get %s - %w", namespace, err)
+	}
+	if ns.Annotations == nil {
+		ns.Annotations = make(map[string]string)
+	}
+	if _, ok := ns.Annotations[namespaceAnnotationNetworkPolicy]; !ok {
```
Comment (brandond) [inline pkg/rke2/np.go:57]: Just to be clear, we don't care about annotation value, just that one exists with this key.

File: pkg/rke2/psp.go (hunk 6 above)
Comment (briandowns) [reply pkg/rke2/psp.go:352]: This only operates in the kube-system namespace.

File: pkg/rke2/np.go (hunk 7 above)
Comment (briandowns) [reply pkg/rke2/np.go:57]: Correct.

File: pkg/rke2/psp.go (hunk 6 above)
Comment (brandond) [reply pkg/rke2/psp.go:352]: OK. Seems a little weird to pass in one NS and operate on another though. Maybe document this behavior in the doc comment?

File: pkg/rke2/np.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+		if _, err := cs.NetworkingV1().NetworkPolicies(namespace).Get(ctx, defaultNetworkPolicyName, metav1.GetOptions{}); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+			if err := cs.NetworkingV1().NetworkPolicies(namespace).Delete(ctx, defaultNetworkPolicyName, metav1.DeleteOptions{}); err != nil {
```
Comment (cjellick) [inline pkg/rke2/np.go:62]: Doesn't this need to go outside the `err != nil` block? At this point, you HAVE gotten an IsNotFound error, so there is nothing to delete. You only want to delete if it was found successfully, ie err == nil

**Pull request merged**